### PR TITLE
Rework mirror service

### LINKED ISSUE #244: Migrate remaining events to use Protobuf schemas
We currently maintain duplicate code for message POJOs in the API server and PoC applications, adding unnecessary overhead, and constantly putting us in danger of various inconsistencies between model classes.

Look into defining message models in a common schema, and automatically generate POJOs from those.

Possible options:

* JSON schema (without schema registry): https://github.com/joelittlejohn/jsonschema2pojo has a Maven plugin to generate POJOs at build time
* JSON schema (with schema registry): https://docs.confluent.io/platform/current/schema-registry/serdes-develop/serdes-json.html
* Avro (requires schema registry): https://docs.confluent.io/platform/current/schema-registry/serdes-develop/serdes-avro.html
* Protobuf (requires schema registry): https://docs.confluent.io/platform/current/schema-registry/serdes-develop/serdes-protobuf.html

For now, I think I'd prefer to *not* introduce the schema registry. Unless Redpanda is used (which [bundles a schema registry](https://redpanda.com/blog/schema-registry)), utilizing a schema registry will mean there's an additional service that needs to be taken care of.

### Progress

- [ ] Mirror
- [x] ~Metrics (#386)~
- [x] Notifications (#415)
- [x] Repo meta analysis (#411)
- [x] Vulnerability analysis (#338)

### LINKED ISSUE #304: Long-running mirroring causes rebalance issues
When mirroring the NVD without an API key, the whole procedure takes way longer than when an API key is provided.

The Kafka Streams thread is effectively blocked while mirroring. During processing, it is not sending heartbeats to the Kafka broker, causing it to be kicked out of the consumer group (broker assumes it's dead):

```
2023-02-01 18:09:40,677 WARN  [org.apa.kaf.str.pro.int.StreamThread] (hyades-mirror-service-fe261319-eda1-44ee-a748-c5f33f51e2f8-StreamThread-1) stream-thread [hyades-mirror-service-fe261319-eda1-44ee-a748-c5f33f51e2f8-StreamThread-1] Detected that the thread is being fenced. This implies that this thread missed a rebalance and dropped out of the consumer group. Will close out all assigned tasks and rejoin the consumer group.: org.apache.kafka.streams.errors.TaskMigratedException: Consumer committing offsets failed, indicating the corresponding thread is no longer part of the group; it means all tasks belonging to this thread should be migrated.
        at org.apache.kafka.streams.processor.internals.TaskExecutor.commitOffsetsOrTransaction(TaskExecutor.java:225)
        at org.apache.kafka.streams.processor.internals.TaskExecutor.commitTasksAndMaybeUpdateCommittableOffsets(TaskExecutor.java:157)
        at org.apache.kafka.streams.processor.internals.TaskManager.commitTasksAndMaybeUpdateCommittableOffsets(TaskManager.java:1118)
        at org.apache.kafka.streams.processor.internals.TaskManager.commit(TaskManager.java:1086)
        at org.apache.kafka.streams.processor.internals.StreamThread.maybeCommit(StreamThread.java:1066)
        at org.apache.kafka.streams.processor.internals.StreamThread.runOnce(StreamThread.java:801)
        at org.apache.kafka.streams.processor.internals.StreamThread.runLoop(StreamThread.java:588)
        at org.apache.kafka.streams.processor.internals.StreamThread.run(StreamThread.java:550)
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:775)
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.posix.thread.PosixPlatformThreads.pthreadStartRoutine(PosixPlatformThreads.java:203)
Caused by: org.apache.kafka.clients.consumer.CommitFailedException: Offset commit cannot be completed since the consumer is not part of an active group for auto partition assignment; it is likely that the consumer was kicked out of the group.
        at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator.sendOffsetCommitRequest(ConsumerCoordinator.java:1311)
        at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator.commitOffsetsSync(ConsumerCoordinator.java:1168)
        at org.apache.kafka.clients.consumer.KafkaConsumer.commitSync(KafkaConsumer.java:1506)
        at org.apache.kafka.clients.consumer.KafkaConsumer.commitSync(KafkaConsumer.java:1453)
        at org.apache.kafka.streams.processor.internals.TaskExecutor.commitOffsetsOrTransaction(TaskExecutor.java:222)
        ... 9 more
```

### LINKED ISSUE #354: Only publish events to `dtrack.vulnerability` when the respective vulnerability actually changed
When mirroring datasources, the mirror-service will currently blindly publish vulnerability data to the `dtrack.vulnerability` topic.

This means that the API server will be under increased pressure whenever a mirror operation is running, as it has to re-ingest >200k events every single time.

It would be a lot more efficient if the mirror-service would keep track of which vulnerabilities it has already seen, and which did not change since it saw them the last time. To reduce load on the API server, the mirror-service should only publish vulnerability events to the topic when the respective vuln is new, or was updated.

### PR DESCRIPTION
**Pull request by nscuro**
Fixes #304 
Closes #354
Closes #244

image

### TODOs

- [x] Tests @nscuro 
- [x] Parsing logic (can be re-used from existing implementation, but must be adapted to use Protobuf models) @VithikaS 
  - https://github.com/DependencyTrack/hyades/blob/main/mirror-service/src/main/java/org/hyades/nvd/NvdToCyclonedxParser.java
  - https://github.com/DependencyTrack/hyades/blob/main/mirror-service/src/main/java/org/hyades/osv/OsvToCyclonedxParser.java
- [x] OSV mirroring @mehab 
- [x] API server changes @sahibamittal 
  - Input topic for mirroring commands have changed; Previously multiple, now only one
  - Parsing has to be changed to Protobuf
  - Account for vulnerabilities from GitHub when parsing

### CODE REVIEW

File: mirror-service-x/src/main/java/org/hyades/vulnmirror/datasource/AbstractDatasourceMirror.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+    /**
+     * Publish a {@link Notification} of group {@link Group#GROUP_DATASOURCE_MIRRORING} to Kafka.
+     *
+     * @param level   The {@link Level} of the {@link Notification}
+     * @param title   The title of the {@link Notification}
+     * @param content The content of the {@link Notification}
+     */
+    protected void dispatchNotification(final Level level, final String title, final String content) {
+        kafkaProducer.send(new ProducerRecord<>(
```
[the remark this answers is not in the document]
Comment (nscuro) [reply mirror-service-x/src/main/java/org/hyades/vulnmirror/datasource/AbstractDatasourceMirror.java:121]: That is intentional.

@sonatype-lift ignore

File: mirror-service-x/src/main/java/org/hyades/vulnmirror/KafkaStreamsTopologyProducer.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+                    try {
+                        return KeyValue.pair(Datasource.valueOf(key), value);
+                    } catch (IllegalArgumentException e) {
+                        return null;
+                    }
+                }, Named.as("parse_datasource_from_key"))
+                .foreach((datasource, value) -> datasourceMirrors.stream()
+                                .filter(mirror -> mirror.supportsDatasource(datasource))
+                                .findAny()
+                                .ifPresent(DatasourceMirror::doMirror),
```
[the remark this answers is not in the document]
Comment (nscuro) [reply mirror-service-x/src/main/java/org/hyades/vulnmirror/KafkaStreamsTopologyProducer.java:62]: That is intentional.

@sonatype-lift ignore

File: mirror-service-x/src/main/java/org/hyades/vulnmirror/datasource/AbstractDatasourceMirror.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (nscuro) [reply mirror-service-x/src/main/java/org/hyades/vulnmirror/datasource/AbstractDatasourceMirror.java:121]: @sonatype-lift ignore

File: mirror-service-x/src/main/java/org/hyades/vulnmirror/KafkaStreamsTopologyProducer.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (nscuro) [reply mirror-service-x/src/main/java/org/hyades/vulnmirror/KafkaStreamsTopologyProducer.java:62]: @sonatype-lift ignore

File: mirror-service-x/src/main/java/org/hyades/vulnmirror/datasource/github/GitHubAdvisoryToCdxParser.java [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+            });
+        }
+        return cwes;
+    }
+
+    private static PackageURL generatePurlFromGitHubVulnerability(final io.github.jeremylong.ghsa.Vulnerability vuln) {
+        final String purlType = mapGitHubEcosystemToPurlType(vuln.getPackage().getEcosystem());
+        try {
+            if (purlType != null) {
+                if (PackageURL.StandardTypes.NPM.equals(purlType) && vuln.getPackage().getName().contains("/")) {
```
[the remark this answers is not in the document]
Comment (nscuro) [reply mirror-service-x/src/main/java/org/hyades/vulnmirror/datasource/github/GitHubAdvisoryToCdxParser.java:197]: @sonatype-lift ignore

File: mirror-service-x/src/main/java/org/hyades/vulnmirror/datasource/github/GitHubAdvisoryToCdxParser.java [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+                    final String[] parts = vuln.getPackage().getName().split("/");
+                    return PackageURLBuilder.aPackageURL().withType(purlType).withNamespace(parts[0]).withName(parts[1]).build();
+                } else if (PackageURL.StandardTypes.MAVEN.equals(purlType) && vuln.getPackage().getName().contains(":")) {
```
[the remark this answers is not in the document]
Comment (nscuro) [reply mirror-service-x/src/main/java/org/hyades/vulnmirror/datasource/github/GitHubAdvisoryToCdxParser.java:200]: @sonatype-lift ignore

File: mirror-service/src/main/java/org/hyades/vulnmirror/datasource/AbstractDatasourceMirror.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+    /**
+     * Publish a {@link Notification} of group {@link Group#GROUP_DATASOURCE_MIRRORING} to Kafka.
+     *
+     * @param level   The {@link Level} of the {@link Notification}
+     * @param title   The title of the {@link Notification}
+     * @param content The content of the {@link Notification}
+     */
+    protected void dispatchNotification(final Level level, final String title, final String content) {
+        kafkaProducer.send(new ProducerRecord<>(
```
[the remark this answers is not in the document]
Comment (nscuro) [reply mirror-service/src/main/java/org/hyades/vulnmirror/datasource/AbstractDatasourceMirror.java:121]: @sonatype-lift ignore

File: mirror-service/src/main/java/org/hyades/vulnmirror/datasource/osv/OsvClient.java [hunk 6]
```diff
@@ -29,19 +27,22 @@
...
-        this.apiBaseUrl = apiBaseUrl.orElse(null);
+        this.osvConfig = osvConfig;
     }
 
     public Path downloadEcosystemZip(String ecosystem) throws IOException {
-        final var request = new HttpGet(this.apiBaseUrl + "/" + URLEncoder.encode(ecosystem, StandardCharsets.UTF_8).replace("+", "%20")
+        if(ecosystem == null){
+            throw new NullPointerException("Ecosystem cannot be null");
+        }
+        final var request = new HttpGet(this.osvConfig.baseurl().get() + "/" + URLEncoder.encode(ecosystem, StandardCharsets.UTF_8).replace("+", "%20")
```
[the remark this answers is not in the document]
Comment (nscuro) [reply mirror-service/src/main/java/org/hyades/vulnmirror/datasource/osv/OsvClient.java:45]: @sonatype-lift ignore

File: mirror-service/src/main/java/org/hyades/vulnmirror/datasource/github/GitHubAdvisoryToCdxParser.java [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+    }
+
+    private static PackageURL generatePurlFromGitHubVulnerability(final io.github.jeremylong.ghsa.Vulnerability vuln) {
+        final String purlType = mapGitHubEcosystemToPurlType(vuln.getPackage().getEcosystem());
+        try {
+            if (purlType != null) {
+                if (PackageURL.StandardTypes.NPM.equals(purlType) && vuln.getPackage().getName().contains("/")) {
+                    final String[] parts = vuln.getPackage().getName().split("/");
+                    return PackageURLBuilder.aPackageURL().withType(purlType).withNamespace(parts[0]).withName(parts[1]).build();
+                } else if (PackageURL.StandardTypes.MAVEN.equals(purlType) && vuln.getPackage().getName().contains(":")) {
```
[the remark this answers is not in the document]
Comment (nscuro) [reply mirror-service/src/main/java/org/hyades/vulnmirror/datasource/github/GitHubAdvisoryToCdxParser.java:200]: @sonatype-lift ignore

File: mirror-service/src/main/java/org/hyades/vulnmirror/datasource/github/GitHubAdvisoryToCdxParser.java [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+            });
+        }
+        return cwes;
+    }
+
+    private static PackageURL generatePurlFromGitHubVulnerability(final io.github.jeremylong.ghsa.Vulnerability vuln) {
+        final String purlType = mapGitHubEcosystemToPurlType(vuln.getPackage().getEcosystem());
+        try {
+            if (purlType != null) {
+                if (PackageURL.StandardTypes.NPM.equals(purlType) && vuln.getPackage().getName().contains("/")) {
```
[the remark this answers is not in the document]
Comment (nscuro) [reply mirror-service/src/main/java/org/hyades/vulnmirror/datasource/github/GitHubAdvisoryToCdxParser.java:197]: @sonatype-lift ignore

File: mirror-service/src/main/java/org/hyades/vulnmirror/datasource/osv/OsvMirror.java [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+            while ((line = reader.readLine()) != null) {
+                out.append(line);
+            }
+            var json = new JSONObject(out.toString());
+            Bom bov = new OsvToCyclonedxParser(this.objectMapper).parse(json);
+            if (bov != null) {
+                publishIfChanged(bov);
+            }
+            zipEntry = zipIn.getNextEntry();
+            reader = new BufferedReader(new InputStreamReader(zipIn));
```
[the remark this answers is not in the document]
Comment (nscuro) [reply mirror-service/src/main/java/org/hyades/vulnmirror/datasource/osv/OsvMirror.java:78]: @sonatype-lift ignore

File: mirror-service/src/main/java/org/hyades/vulnmirror/datasource/osv/OsvClient.java [hunk 10]
```diff
@@ -17,39 +16,37 @@
...
-        this.objectMapper = objectMapper;
-        this.apiBaseUrl = apiBaseUrl.orElse(null);
+        this.osvConfig = osvConfig;
     }
 
     public Path downloadEcosystemZip(String ecosystem) throws IOException {
-        final var request = new HttpGet(this.apiBaseUrl + "/" + URLEncoder.encode(ecosystem, StandardCharsets.UTF_8).replace("+", "%20")
+        final String baseUrl = osvConfig.baseUrl().orElseThrow(() -> new IllegalStateException("No base URL configured"));
+        ecosystem = Objects.requireNonNull(StringUtils.trimToNull(ecosystem), "Ecosystem cannot be null");
+        final var request = new HttpGet(baseUrl + "/" + URLEncoder.encode(ecosystem, StandardCharsets.UTF_8).replace("+", "%20")
```
[the remark this answers is not in the document]
Comment (nscuro) [reply mirror-service/src/main/java/org/hyades/vulnmirror/datasource/osv/OsvClient.java:41]: @sonatype-lift ignore

**Comment by nscuro** [comment]
Coverage is not as high as it should be because we exercise large chunks of the code in integration tests, which are executed against binary artifacts (e.g. the JAR or native image).

**Review by mehab — APPROVED** [review]
lgtm

**Review by sahibamittal — APPROVED** [review]

**Pull request merged**